forem/forem
# Tag autocomplete: storybook multi-select autocomplete component

## What type of PR is this? (check all applicable)


## Description

Here's a wee loom showing the functionality expected: https://www.loom.com/share/738a2f879340495dbf794406ccef1dd7

This PR addresses the first step of the tag autocomplete epic, by adding a multi-select autocomplete component to Storybook. At this time, the component is totally generic, and I'll later address the work to show a richer UI for the tag suggestions/selections (and actually add it to the editor).

## Related Tickets & Documents

Closes https://github.com/forem/forem/issues/15421
Contributes to epic https://github.com/forem/forem/issues/14845

## QA Instructions, Screenshots, Recordings

Check out the loom video linked above for a rundown of what's expected, but as a quick checklist:

- As a user types, the displayed options should update (NB: in Storybook, the options available are the numbers one through ten)
- User should be able to select an item from multiple options in a list by either keyboard or mouse
- On selection, the item should appear with an 'X' icon next to it
- Clicking the 'X' should remove the selection
- Clicking on the selection's text, rather than the 'X' should place it back in 'edit mode'
- On selecting an item, text should be cleared and a new selection should be possible by continuing to type
- Typing a comma or space, should have the effect of selecting the currently entered text
- Pressing delete should place the last selection (if any) into 'edit mode', and if you keep pressing delete until there's no text left, it should delete the option altogether

### UI accessibility concerns?

There's a fair bit of complexity in the component, and there isn't a standard authoring practice for a combobox with multiple selections, edit mode, etc. In saying that though, the following features exist:

- The core functionality - type to search, select from a list - conforms to the [combobox authoring practices](https://www.w3.org/TR/wai-aria-practices-1.1/#combobox). We can be confident this meets user expectations.
- There is a hidden list of the current selections which is updated every time a selection is removed or added. This hidden list can also be read in 'reading' mode by screen reader users.
- The edit/delete buttons are semantically grouped together, and have aria labels describing their actions
- The edit/delete buttons are keyboard accessible, and grouped inside the combobox to help convey their relation to the 'autocomplete' functionality. This is inspired partly by [the research and implementation described in Sarah Higley's article](https://www.24a11y.com/2019/select-your-poison-part-2/) in which they describe user testing this pattern. This makes me feel somewhat confident in the accessibility of this.

We have a clean bill of health from `axe` and the Storybook a11y checker.

## Added/updated tests?

- [X] No, and this is why: This component is still somewhat subject to change, and I'd rather write these tests once we start using the component for the tags autocomplete, by which point any necessary changes will be flushed out

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

      [Storybook](https://storybook.forem.com/) (for Crayons components)
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
      or in a [forem.dev](http://forem.dev) post
- [X] This change does not need to be communicated, and this is why not: it represents a work in progress

## [optional] Are there any post deployment tasks we need to perform?

N/A

## [optional] What gif best describes this PR or how it makes you feel?

![It's going to be big!](https://media.giphy.com/media/ZC3ro4RBCJmhGERE8T/giphy.gif)
